addons/website/static/src/builder/plugins/customize_website_plugin.js

```diff
@@ -723,36 +727,24 @@ export class PreviewableWebsiteConfigAction extends BuilderAction {
         }
         return el.classList.contains(params.previewClass);
     }
-    apply({ editingElement: el, isPreviewing, params, selectableContext }) {
+    apply({ editingElement: el, isPreviewing, params }) {
         if (params.previewClass) {
             el.classList.add(params.previewClass);
         }
         if (!isPreviewing) {
-            const applyViewAction = (reverting) => {
-                const prepareRecord = (record, disable) => {
-                    if (record.startsWith("!")) {
-                        const recordKey = record.substring(1);
-                        this.dependencies.customizeWebsite.setViewOnSave(recordKey, !disable);
-                    } else {
-                        this.dependencies.customizeWebsite.setViewOnSave(record, disable);
-                    }
-                };
-                const records = params["views"] || [];
-                if (selectableContext) {
-                    for (const item of selectableContext.items) {
-                        for (const a of item.getActions()) {
-                            if (a.actionId === "previewableWebsiteConfig") {
-                                for (const record of a.actionParam["views"] || []) {
-                                    // Start by disabling them all...
-                                    prepareRecord(record, !reverting);
-                                }
-                            }
-                        }
-                    }
-                    for (const record of records) {
-                        // ...then enable the selected one
-                        prepareRecord(record, reverting);
+            const viewsToApply = params["views"] || [];
+            const applyOperations = viewsToApply.map((view) => ({
+                view: view,
+                wasToDisable: this.dependencies.customizeWebsite.getViewsOnSave(view).toDisable,
+            }));
+            const applyViewAction = (is_undoing) => {
+                for (const op of applyOperations) {
+                    if (op.wasToDisable) {
+                        // If the view was previously added to the "toDisable"
+                        // list, we remove it from there.
+                        this.processView(op.view, false, !is_undoing);
```

Comment:
Why not manage this stuff in setViewsOnSave ? (if op.wasToDisable)
I think your action shouldn't care about it (already disabled etc).

processView and setViewsOnSave should review a list of view.
For the undo,  setViewsOnSave could maybe return a callback that will undo.  And you can use this function in the reverse maybe ?